src/main/java/edu/ucsd/sbrg/escher/converter/Escher2SBML.java

```diff
@@ -63,1048 +54,1265 @@
         } else {
-          isProduct = true;
+            logger.warning(format(bundle.getString("Escher2SBML.undefinedID"),
+                    node.toString()));
         }
-      }
     }
-      if (srGlyph != null) {
-        LineSegment ls = convertSegment(segment, escherMap, srGlyph.createCurve(), xOffset, yOffset);
-        // memorize the toNode id of the segment in the lineSegment to make access easier later on.
-        ls.putUserObject(ESCHER_NODE_LINK,
-          isProduct ? segment.getFromNodeId() : segment.getToNodeId());
-      } else {
-        segments.add(segment);
-      }
+
+
+    /**
+     * Converts the text label nodes to text glyphs
+     *
+     * @param node    The text label node to be converted
+     * @param layout  The {@link Layout} object of the SBML model
+     * @param xOffset x-offset of the document
+     * @param yOffset y-offset of the document
+     */
+    private void convertTextLabel(Node node, Layout layout, double xOffset,
+                                  double yOffset) {
+        createTextGlyph(node, layout, xOffset, yOffset);
+        logger.info(format(bundle.getString("Escher2SBML.skippingNode"), node.toString()));
     }
-    Set<Segment> done = new HashSet<Segment>();
-    for (SpeciesReferenceGlyph srGlyph : srgMap.values()) {
-      Curve curve = srGlyph.getCurve();
-      Node toNode = escherMap.getNode( curve.getCurveSegment(curve.getCurveSegmentCount() - 1).getUserObject(ESCHER_NODE_LINK).toString());
-      while (!toNode.isMidmarker()) {
-        for (Segment segment : segments) {
-          if (tryToAttach(segment, curve, curve.getCurveSegmentCount() - 1,
-            escherMap, xOffset, yOffset)) {
-            done.add(segment);
-            toNode = escherMap.getNode(curve.getCurveSegment(curve.getCurveSegmentCount() - 1).getUserObject(ESCHER_NODE_LINK).toString());
-            if (toNode.isMidmarker()) {
-              break;
+
+
+    /**
+     * Converts an {@link EscherReaction} to a {@link ReactionGlyph}
+     *
+     * @param escherReaction The reaction to be converted
+     * @param escherMap      The {@link EscherMap} the {@code escherReaction} is in
+     * @param layout         The {@link Layout} object of the SBML model
+     * @param node2glyph     A hash map of the node ids and their converted glyph ids
+     * @param xOffset        x-offset of the document
+     * @param yOffset        y-offset of the document
+     */
+    private Reaction convertReaction(EscherReaction escherReaction,
+                                     EscherMap escherMap, Layout layout, Map<String, String> node2glyph,
+                                     double xOffset, double yOffset) {
+        Metabolite met = escherReaction.getMetaboliteList().get(0);
+        Node metNode = escherMap.getNode(met.getNodeRefId());
+        boolean isExchange = (escherReaction.getMetaboliteCount() == 1);
+        String exchangeNodeId = metNode.getId() + "_ex";
```

Comment:
Since these identifier additions re-occur in multiple places, it could also make sense to define them in an `interface` as static constant fields. Next, classes needing these constants can simply implement that interface and inherit all the defined `String` literals.